vertx-web-session-stores/vertx-web-sstore-caffeine/src/main/java/io/vertx/ext/web/sstore/caffeine/impl/CaffeineSessionStoreImpl.java

```diff
@@ -0,0 +1,131 @@
+package io.vertx.ext.web.sstore.caffeine.impl;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import io.vertx.codegen.annotations.Nullable;
+import io.vertx.core.Closeable;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.internal.CloseFuture;
+import io.vertx.core.internal.ContextInternal;
+import io.vertx.core.internal.VertxInternal;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.auth.prng.VertxContextPRNG;
+import io.vertx.ext.web.Session;
+import io.vertx.ext.web.sstore.AbstractSession;
+import io.vertx.ext.web.sstore.SessionStore;
+import io.vertx.ext.web.sstore.caffeine.CaffeineSessionStore;
+import io.vertx.ext.web.sstore.impl.SharedDataSessionImpl;
+
+import java.time.Duration;
+
+/**
+ * @author <a href="mailto:lazarbulic@gmail.com">Lazar Bulic</a>
+ */
+public class CaffeineSessionStoreImpl implements SessionStore, CaffeineSessionStore {
+
+  /**
+   * Default name for the shared Caffeine sessions cache.
+   */
+  private static final String DEFAULT_SESSION_CACHE_NAME = "vertx-web.caffeine.sessions";
+
+
+  private Cache<String, Session> localCaffeineCache;
+  private VertxContextPRNG random;
+  private String sessionCacheName;
+  private Closeable closeable;
+
+  private VertxInternal vertx;
+
+  public CaffeineSessionStoreImpl() {
+    // required for the service loader
+  }
+
+  @Override
+  public Session createSession(long timeout) {
+    return new SharedDataSessionImpl(random, timeout, DEFAULT_SESSIONID_LENGTH);
+  }
+
+  @Override
+  public Session createSession(long timeout, int length) {
+    return new SharedDataSessionImpl(random, timeout, length);
+  }
+
+  @Override
+  public SessionStore init(Vertx vertx, JsonObject options) {
+    // initialize a secure random
+    this.random = VertxContextPRNG.current(vertx);
+    this.vertx = (VertxInternal) vertx;
+    this.sessionCacheName = options.getString("cacheName", DEFAULT_SESSION_CACHE_NAME);
+    final ContextInternal ctx = ((VertxInternal) vertx).getOrCreateContext();
+    CloseFuture closeFuture = new CloseFuture();
+    localCaffeineCache = this.vertx.createSharedResource("__vertx.shared.caffeine.sessions.store", sessionCacheName, closeFuture, cf_ -> {
+      Cache<String, Session> localCaffeineCache = Caffeine.newBuilder()
+        .executor(cmd -> ctx.runOnContext(v -> cmd.run()))
```

Comment:
```suggestion
```

I believe this is only needed for loading tasks or notifications, which we don't use.

Reply:
I have considered removing this also (same story sync vs async) but then I read this:
https://github.com/ben-manes/caffeine/blob/701df97f7597ffec5ce772014ed43b963e56a453/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Caffeine.java#L329-L347


There are maintenance task that will be run via this context, probably eviction so I left this so that it is executed on already existing event loop instead of using default ForkJoinPool.

Reply:
Good point, thanks